CRM/Mailing/BAO/Mailing.php

```diff
@@ -109,46 +109,227 @@ public static function getRecipientsCount($job_id, $mailing_id = NULL) {
   }
 
   /**
-   * note that $job_id is used only as a variable in the temp table construction
-   * and does not play a role in the queries generated.
-   * @param int $job_id
-   *   (misnomer) a nonce value used to name temporary tables.
-   * @param int $mailing_id
-   * @param bool $storeRecipients
-   * @param bool $dedupeEmail
-   * @param null $mode
+   * This function is created after simplifying the code of self::getRecipients(..) and these are the changes made
```

Comment:
I would only put permanent documentation in the PR, and put descriptions of changes / improvements that are relevant while reviewing into comments on the PR. That way, the repo only has comments that continue to be relevant. So, for example, the PR should contain:
This function provides the recipients of a CiviMail mailing.
It provides a simplified implementation of an earlier function, now removed, self::getRecipients(..). 
NB: The Smart Group contact cache is refreshed immediately by this call for all included and excluded smart groups in the mailing. 

In comments on the PR, put in 1 - 5b.

Now, a question: do we really want this code to override the smart cache timeout on groups and force their refresh? Wouldn't it be better to rely on the cache than to force it to be rebuilt? We could provide an additional button for those rare cases where someone is concerned about forcing a rebuild, and willing to accept the possibly long rebuild time. Recall that this can take 5 - 10 minutes for some large complex hierarchical smart groups.

Reply:
Actually, we are not forcing to rebuilt group contact cache because          ```CRM_Contact_BAO_GroupContactCache::load($groupDAO)``` respects the cache and refresh time, and on basis of it which it will decide to rebuild contact cache for smart group identified by ```$groupDAO``` which is how it's been done in other places.